design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,301 @@
+### Inter-service communications
+This documents describes how exactly services will be communicating in a distributed setup.
+#### Service declaration
+Services will be declared in Api modules using interfaces. These modules will be installed on all nodes.
+Example:
+```php
+namespace Magento\ServiceBApi;
+
+interface BManagerInterface
+{
+    public function save(BDataInterface $data): BDataInterface;
+    
+    public function delete(string $id): void;
+}
+```
+or for DTOs
+```php
+namespace Magento\ServiceBApi\Data;
+
+interface BDataInterface
+{
+    public function getId(): string;
+  
+    public function getName(): string;
+}
+```
+The API modules will just contain these declarations and config files (like acl.xml, config.xml) describing module
+without describing actual implementation.
+ 
+Even services that are not written using Magento (or even PHP) will be declared in this way for nodes using Magento.
+ 
+#### Publicly available Web APIs
+To enable publicly available APIs (that frontend and 3rd party systems can use) the _front node_ will have to have -Module-Webapi
+or -Module-Graphql modules installed.
+ 
+#### Real and proxy modules
+If a node wants to use services locally then it will just have real modules installed (e.g. "Catalog"), if a node is supposed
+to access services remotely it will have proxy modules installed (e.g. "CatalogProxy").
+ 
+Proxy modules will contain services/DTOs implementations and a di.xml file to declare preferences for proxy implementations to be
```

Comment:
DTO implementations should be in *API modules.

Reply:
We would have to rewrite every module. DTOs in API modules are just declarations of data types this module's service contracts accept/return, other modules should not care what the actual implementations are. Plus we allow some services to be implemented using different programming languages - implementations for DTOs for those service contracts in PHP would be redundant.